lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,244 @@
+    }
+
+    if (options.signal !== undefined) {
+      validateAbortSignal(options.signal, 'options.signal');
+    }
+
+    if (options.encoding !== undefined) {
+      if (typeof options.encoding !== 'string') {
+        throw new ERR_INVALID_ARG_VALUE('encoding', 'options.encoding');
+      }
+    }
+
+    this.#options = options;
+  }
+
+  close() {
+    const { unwatchFile } = lazyLoadFsSync();
+    this.#closed = true;
+
+    for (const file of this.#files.keys()) {
+      unwatchFile(file);
+    }
+
+    this.#files.clear();
+    this.emit('close');
+  }
+
+  /**
+   * @param {string} file
+   * @return {string}
+   */
+  #getPath(file) {
+    validateString(file, 'file');
+
+    if (file === this.#rootPath) {
+      return this.#rootPath;
+    }
+
+    return path.relative(this.#rootPath, file);
+  }
+
+  #unwatchFolder(file) {
+    validateString(file, 'file');
+
+    const { unwatchFile } = lazyLoadFsSync();
+
+    for (const filename of this.#files.keys()) {
+      if (StringPrototypeStartsWith(filename, file)) {
+        unwatchFile(filename);
+      }
+    }
+  }
+
+  async #watchFolder(folder) {
+    validateString(folder, 'folder');
+
+    const { opendir } = lazyLoadFsPromises();
+
+    try {
+      const files = await opendir(folder);
+
+      for await (const file of files) {
+        const f = path.join(folder, file.name);
+
+        if (this.#closed) {
+          break;
+        }
+
+        if (!this.#files.has(f)) {
+          this.#files.set(f, file);
+          this.emit('change', 'rename', this.#getPath(f));
+
+          if (file.isDirectory()) {
+            await this.#watchFolder(f);
+          } else {
+            this.#watchFile(f);
+          }
+        }
+      }
+    } catch (error) {
+      this.emit('error', error);
+    }
+  }
+
+  #watchFile(file) {
+    validateString(file, 'file');
+
+    const { watchFile } = lazyLoadFsSync();
+
+    if (this.#closed) {
+      return;
+    }
+
+    const existingStat = this.#files.get(file);
+
+    watchFile(file, {
+      persistent: this.#options.persistent,
+    }, (statWatcher, previousStatWatcher) => {
+      if (existingStat && !existingStat.isDirectory() &&
+        statWatcher.nlink !== 0 && existingStat.mtime?.getTime() === statWatcher.mtime?.getTime()) {
+        return;
+      }
+
+      this.#files.set(file, statWatcher);
+
+      if (statWatcher.birthtimeMs === 0 && previousStatWatcher.birthtimeMs !== 0) {
+        // The file is now deleted
+        this.#files.delete(file);
+        this.emit('change', 'rename', this.#getPath(file));
+
+        if (statWatcher.isDirectory()) {
+          this.#unwatchFolder(file);
+        }
+      } else if (statWatcher.isDirectory()) {
+        this.emit('change', 'change', this.#getPath(file));
+        this.#watchFolder(file);
+      } else {
+        this.emit('change', 'change', this.#getPath(file));
+      }
+    });
+  }
+
+  [kFSWatchStart](filename) {
+    validateString(filename, 'filename');
```

Comment:
Can't `filename` be a WHATWG `URL` instance? IMO we should let `statSync` do the validation.